src/mjolnir/graphenhancer.cc

```diff
@@ -94,58 +96,234 @@ struct enhancer_stats {
+    // no special uses
+    std::array<uint32_t, 8> way;
+    // ramps
+    std::array<uint32_t, 6> link_exiting;
+    // turn channel
+    std::array<uint32_t, 6> link_turning;
+    // roundabout
+    std::array<uint32_t, 8> roundabout;
+    // driveway, alley, parking_aisle, drive-through
+    std::array<uint32_t, 4> service;
+
+    template <typename container_t>
+    void parse(const rapidjson::Value& obj, const std::string& name, container_t& entries) {
+      auto arr = obj[name].GetArray();
+      if (arr.Size() != entries.size())
+        throw std::runtime_error(name + " must have " + std::to_string(entries.size()) + " speeds");
+      size_t i = 0;
+      for (const auto& speed : arr) {
+        entries[i++] = speed.GetUint();
       }
     }
-    directededge.set_speed(speed);
 
-    // Done processing links so return...
-    return;
+    SpeedTable(const rapidjson::Value& obj) {
+      parse(obj, "way", way);
+      parse(obj, "link_exiting", link_exiting);
+      parse(obj, "link_turning", link_turning);
+      parse(obj, "roundabout", roundabout);
+      service[0] = obj["driveway"].GetUint();
+      service[1] = obj["alley"].GetUint();
+      service[2] = obj["parking_aisle"].GetUint();
+      service[3] = obj["drive-through"].GetUint();
+    }
+  };
+
+  // 2 letter country and 2 letter state as key with urban and rural speed tables as value
+  std::unordered_map<std::string, std::array<SpeedTable, 2>> tables;
+
+  SpeedAssigner(const boost::optional<std::string>& config_file) {
+    if (!config_file)
+      return;
+    try {
+      auto doc = rapidjson::read_json(*config_file);
+      if (doc.HasParseError())
+        throw std::runtime_error("malformed json");
+      if (!doc.IsArray())
+        throw std::runtime_error("must be a json array");
+
+      // loop over each country/state pair
+      for (const auto& cs : doc.GetArray()) {
+        std::string code = cs["iso3166-1"].GetString();
+        code += cs["iso3166-2"].GetString();
+        tables.emplace(std::move(code), std::array<SpeedTable, 2>{
+                                            SpeedTable(cs["urban"]),
+                                            SpeedTable(cs["rural"]),
+                                        });
+      }
+    } // something went wrong with parsing or opening the file
+    catch (const std::exception& e) {
+      LOG_WARN(std::string("Could not load default speeds: ") + e.what());
+      tables.clear();
+    } // something else was thrown
+    catch (...) {
+      LOG_WARN("Could not load default speeds from config");
+      tables.clear();
+    }
   }
 
-  // If speed is assigned from an OSM max_speed tag we only update it based
-  // on surface type.
-  if (directededge.speed_type() == SpeedType::kTagged) {
-    // Reduce speed on rough pavements. TODO - do we want to increase
-    // more on worse surface types?
-    if (directededge.surface() >= Surface::kPavedRough) {
+  bool FromConfig(DirectedEdge& directededge,
```

Comment:
this is the main function that either uses a speed from config or aborts and relies on the previous method. anywhere where the method returns `false` it is signalling that it couldnt assign a speed for this edge. ive commented the sections below to make it pretty clear. if its a kind of edge we know we cant find a speed for either becuase its a special edge that we dont handle or because the geography is not in the config then we abort immediately and rely on the conventional speed setting. once it gets past that point it figures out whether it needs rural or urban speeds. finally it picks which speed to assign.

ill add a javadoc to this function